internal/node-patches/src/fs.ts

```diff
@@ -0,0 +1,479 @@
+  const origReadlinkSync = fs.readlinkSync.bind(fs);
+  const origReaddir = fs.readdir.bind(fs);
+  const origReaddirSync = fs.readdirSync.bind(fs);
+
+  const { isEscape, isOutPath } = escapeFunction(root);
+
+  const logged: { [k: string]: boolean } = {};
+
+  //tslint:disable-next-line:no-any
+  fs.lstat = (...args: any[]) => {
+    const ekey = new Error('').stack || '';
+    if (!logged[ekey]) {
+      logged[ekey] = true;
+    }
+
+    let cb = args.length > 1 ? args[args.length - 1] : undefined;
+    // preserve error when calling function without required callback.
+    if (cb) {
+      cb = once(cb);
+      args[args.length - 1] = (err: Error, stats: Stats) => {
+        if (err) return cb(err);
+
+        const linkPath = path.resolve(args[0]);
+        // if this is not a symlink or the path is not inside the root it has no way to escape.
+        if (!stats.isSymbolicLink() || !root || isOutPath(linkPath)) {
+          return cb(null, stats);
+        }
+
+        return origReadlink(
+          args[0],
+          (err: Error & { code: string }, str: string) => {
+            if (err) {
+              if (err.code === 'ENOENT') {
+                return cb(false, stats);
+              } else if (err.code === 'EINVAL') {
+                // readlink only returns einval when the target is not a link.
+                // so if we found a link and it's no longer a link someone raced file system modifications.
+                // we return the error but a strong case could be made to return the original stat.
+                return cb(err);
+              } else {
+                // some other file system related error.
+                return cb(err);
+              }
+            }
+
+            str = path.resolve(path.dirname(args[0]), str);
+
+            if (isEscape(str, args[0])) {
+              // if it's an out link we have to return the original stat.
+              return origStat(
+                args[0],
+                (err: Error & { code: string }, plainStat: Stats) => {
+                  if (err && err.code === 'ENOENT') {
+                    //broken symlink. return link stats.
+                    return cb(null, stats);
+                  }
+                  cb(err, plainStat);
+                }
+              );
+            }
+            // its a symlink and its inside of the root.
+            cb(false, stats);
+          }
+        );
+      };
+    }
+    origLstat(...args);
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.realpath = (...args: any[]) => {
+    let cb = args.length > 1 ? args[args.length - 1] : undefined;
+    if (cb) {
+      cb = once(cb);
+      args[args.length - 1] = (err: Error, str: string) => {
+        if (err) return cb(err);
+        if (isEscape(str, args[0])) {
+          cb(false, path.resolve(args[0]));
+        } else {
+          cb(false, str);
+        }
+      };
+    }
+    origRealpath(...args);
+  };
+
+  //tslint:disable-next-line:no-any
```

Comment:
nit: usually `any` isn't actually needed. If we don't care what the args are maybe `Array<{}>`?

Reply:
the real @types/node use any quite a bit and they don't leak because they're monkey patches. So these types only impact the development experience for other authors of this package and Array<{}> doesn't give you anything as far as intelisense but hint incorrectly that this library has any say in the typing

Reply:
I think there is a way to use `typeof` to just say "this should be the same type as the function in the node fs module" but I don't care enough to try to track that down